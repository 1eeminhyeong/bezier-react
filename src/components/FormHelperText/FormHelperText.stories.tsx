/* External dependencies */
import React from 'react'
import base from 'paths.macro'
import type { Story, Meta } from '@storybook/react'

/* Internal dependencies */
import { getTitle } from '../../utils/storyUtils'
import FormHelperText from './FormHelperText'
import FormHelperTextProps from './FormHelperText.types'

export default {
  title: getTitle(base),
  component: FormHelperText,
} as Meta

const Template: Story<FormHelperTextProps> = props => <FormHelperText {...props} />

export const Primary: Story<FormHelperTextProps> = Template.bind({})
Primary.args = {
  id: 'test',
  errorMessage: '',
  children: 'Description',
}
