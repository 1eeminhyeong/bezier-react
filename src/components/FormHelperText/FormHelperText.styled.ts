/* Internal dependencies */
import { styled } from '../../foundation'
import { Text } from '../Text'

export const HelperText = styled(Text)`
  padding: 0 2px;
`
