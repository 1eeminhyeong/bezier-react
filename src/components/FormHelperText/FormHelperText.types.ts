/* Internal dependencies */
import { IdentifierProps, ChildrenComponentProps } from '../../types/ComponentProps'

export default interface FormHelperTextProps extends ChildrenComponentProps, IdentifierProps {
  hasError?: boolean
}
