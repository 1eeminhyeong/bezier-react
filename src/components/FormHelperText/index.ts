export { default as FormHelperText } from './FormHelperText'
export type { default as FormHelperTextProps } from './FormHelperText.types'
