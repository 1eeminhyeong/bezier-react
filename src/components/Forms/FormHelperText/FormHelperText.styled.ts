/* Internal dependencies */
import { styled } from 'Foundation'
import { Text } from 'Components/Text'

export const HelperText = styled(Text)`
  display: block;
  /* FIXME: padding 스타일링을 text prop을 통해서 할 수 있도록 개선 */
  padding: 0 2px;
  text-align: left;
`
