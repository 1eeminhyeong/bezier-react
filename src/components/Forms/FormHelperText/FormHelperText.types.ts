/* Internal dependencies */
import { BezierComponentProps, ChildrenProps, IdentifierProps } from 'Types/ComponentProps'
import { TextProps } from 'Components/Text'
import { FormComponentProps } from 'Components/Forms/Form.types'

export default interface FormHelperTextProps extends
  BezierComponentProps,
  ChildrenProps,
  Partial<IdentifierProps>,
  FormComponentProps,
  TextProps {}
