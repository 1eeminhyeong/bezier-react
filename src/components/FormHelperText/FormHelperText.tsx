/* External dependencies */
import React, { forwardRef, useMemo } from 'react'
import { isEmpty } from 'lodash-es'

/* Internal dependencies */
import { Typography, SemanticNames } from '../../foundation'
import type FormHelperTextProps from './FormHelperText.types'
import * as Styled from './FormHelperText.styled'

export const FORM_HELPER_TEXT_TEST_ID = 'bezier-react-form-helper-text'

function FormHelperText({
  id,
  testId = FORM_HELPER_TEXT_TEST_ID,
  as = 'p',
  hasError,
  children,
  ...rest
}: FormHelperTextProps,
forwardedRef: React.Ref<HTMLParamElement>,
) {
  const color = useMemo<SemanticNames>(() => {
    if (hasError) {
      return 'bgtxt-orange-normal'
    }
    return 'txt-black-dark'
  }, [hasError])

  if (isEmpty(children)) { return null }

  return (
    <Styled.HelperText
      {...rest}
      id={id}
      testId={testId}
      ref={forwardedRef}
      forwardedAs={as}
      marginTop={4}
      typo={Typography.Size13}
      color={color}
    >
      { children }
    </Styled.HelperText>
  )
}

export default forwardRef(FormHelperText)
