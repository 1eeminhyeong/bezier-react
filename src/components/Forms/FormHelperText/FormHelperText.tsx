/* External dependencies */
import React, { forwardRef, useMemo, useCallback } from 'react'
import { isEmpty } from 'lodash-es'

/* Internal dependencies */
import { Typography } from 'Foundation'
import useMergeRefs from 'Hooks/useMergeRefs'
import useFormControlContext from 'Components/Forms/useFormControlContext'
import type { TextProps } from 'Components/Text'
import type FormHelperTextProps from './FormHelperText.types'
import * as Styled from './FormHelperText.styled'

export const FORM_HELPER_TEXT_TEST_ID = 'bezier-react-form-helper-text'

function FormHelperText({
  id,
  testId = FORM_HELPER_TEXT_TEST_ID,
  as = 'p',
  hasError: hasErrorProps,
  typo = Typography.Size13,
  marginTop = 4,
  children,
  ...rest
}: FormHelperTextProps,
forwardedRef: React.Ref<HTMLParamElement>,
) {
  const contextValue = useFormControlContext()

  const setIsRendered = useCallback((node: HTMLElement) => {
    if (!node) { return }
    contextValue?.setHasHelperText(true)
  }, [contextValue])

  const mergedRef = useMergeRefs<HTMLElement>(setIsRendered, forwardedRef)

  const mergedProps = useMemo<TextProps>(() => {
    const hasError = hasErrorProps ?? contextValue?.hasError
    return {
      id: id ?? contextValue?.helperTextId,
      color: hasError
        ? 'bgtxt-orange-normal'
        : 'txt-black-dark',
      'aria-live': hasError ? 'polite' : undefined,
    }
  }, [
    id,
    hasErrorProps,
    contextValue?.helperTextId,
    contextValue?.hasError,
  ])

  if (isEmpty(children)) { return null }

  return (
    <Styled.HelperText
      {...rest}
      {...mergedProps}
      testId={testId}
      ref={mergedRef}
      forwardedAs={as}
      marginTop={marginTop}
      typo={typo}
    >
      { children }
    </Styled.HelperText>
  )
}

export default forwardRef(FormHelperText)
