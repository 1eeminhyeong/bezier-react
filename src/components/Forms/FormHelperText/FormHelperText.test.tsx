/* External dependencies */
import React from 'react'

/* Internal dependencies */
import { LightFoundation } from 'Foundation'
import { render } from 'Utils/testUtils'
import FormHelperText, { FORM_HELPER_TEXT_TEST_ID } from './FormHelperText'
import type FormHelperTextProps from './FormHelperText.types'

describe('FormHelperText >', () => {
  let props: FormHelperTextProps
  const text = 'Lorem ipsum'

  beforeEach(() => {
    props = {
      id: 'test',
      hasError: false,
      children: text,
    }
  })

  const renderFormHelperText = (otherProps?: Partial<FormHelperTextProps>) =>
    render(<FormHelperText {...props} {...otherProps} />)

  it('Snapshot >', () => {
    const { getByTestId } = renderFormHelperText()
    const helperText = getByTestId(FORM_HELPER_TEXT_TEST_ID)

    expect(helperText).toMatchSnapshot()
  })

  it('renders text with correct style when children prop is not empty', () => {
    const { getByTestId } = renderFormHelperText()
    const helperText = getByTestId(FORM_HELPER_TEXT_TEST_ID)

    expect(helperText).toHaveStyle(`color: ${LightFoundation.theme['txt-black-dark']}`)
  })

  it('renders error style text with correct style when hasError prop is true', () => {
    const { getByTestId } = renderFormHelperText({ hasError: true })
    const helperText = getByTestId(FORM_HELPER_TEXT_TEST_ID)

    expect(helperText).toHaveStyle(`color: ${LightFoundation.theme['bgtxt-orange-normal']}`)
  })

  it('renders nothing when children prop is empty', () => {
    const { queryByTestId } = renderFormHelperText({ children: '' })
    const helperText = queryByTestId(FORM_HELPER_TEXT_TEST_ID)

    expect(helperText).toBeNull()
  })
})
