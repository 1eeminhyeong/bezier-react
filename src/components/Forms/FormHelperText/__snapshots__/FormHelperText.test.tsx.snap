// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormHelperText > Snapshot > 1`] = `
.c0 {
  font-size: 13px;
  line-height: 18px;
  margin: 4px 0px 0px 0px;
  font-style: normal;
  font-weight: normal;
  color: #00000066;
  -webkit-transition-delay: 0ms;
  transition-delay: 0ms;
  -webkit-transition-timing-function: cubic-bezier(.3,0,0,1);
  transition-timing-function: cubic-bezier(.3,0,0,1);
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  -webkit-transition-property: color;
  transition-property: color;
}

.c1 {
  display: block;
  padding: 0 2px;
  text-align: left;
}

<p
  class="c0 c1"
  color="txt-black-dark"
  data-testid="bezier-react-form-helper-text"
  id="test"
>
  Lorem ipsum
</p>
`;
